File: .github/scripts/fetch_quant.js
#!/usr/bin/env node
const https = require('https');
const fs = require('fs');

function get(url) {
  return new Promise((resolve, reject) => {
    https.get(url, { timeout: 10000, headers: { 'User-Agent': 'Mozilla/5.0' } }, res => {
      let d = '';
      res.on('data', c => d += c);
      res.on('end', () => resolve(d));
    }).on('error', reject);
  });
}

const FIELDS = 'f2,f3,f4,f5,f6,f7,f8,f9,f10,f12,f14,f15,f16,f17,f18,f20,f21,f23,f25,f37,f38,f62,f115,f152,f168,f169,f170,f171';

async function fetchStocks(page, pz) {
  const url = 'https://push2.eastmoney.com/api/qt/clist/get?pn='+page+'&pz='+pz+'&po=1&np=1&ut=bd1d9ddb04089700cf9c27f6f7426281&fltt=2&invt=2&fid=f62&fs=m:0+t:6,m:0+t:80,m:1+t:2,m:1+t:23&fields='+FIELDS;
  try {
    const text = await get(url);
    const j = JSON.parse(text);
    return { stocks: j.data ? j.data.diff || [] : [], total: j.data ? j.data.total || 0 : 0 };
  } catch(e) { return { stocks: [], total: 0 }; }
}

function pick(stocks) {
  const candidates = [];
  for (const s of stocks) {
    const name = s.f14 || '';
    const code = s.f12 || '';
    if (name.indexOf('ST') >= 0 || name.indexOf('退') >= 0 || name.indexOf('N') >= 0) continue;
    const chg = s.f3 || 0;
    if (chg < 1.0 || chg > 18.0) continue;
    const volRatio = s.f23 || 0;
    if (volRatio < 0.8) continue;
    const turnover = s.f38 || 0;
    if (turnover < 0.5) continue;
    const pe = s.f115 || 0;
    if (pe && pe > 500) continue;
    const price = s.f2 || 0;
    const amt = (s.f62 || 0) / 1e8;
    const circCap = (s.f21 || 1) / 1e10;
    const realTurnover = turnover > 100 ? turnover / 1e8 : turnover;
    let score = 0;
    score += Math.min(chg / 5, 1.5) * 20;
    score += Math.min(volRatio / 3, 1.5) * 20;
    score += Math.min(realTurnover / 5, 1.0) * 15;
    score += Math.min(amt / 5, 1.0) * 15;
    score += Math.min(circCap / 30, 1.0) * 15;
    if (pe && pe > 10 && pe < 50) score += 15;
    const turnDisplay = turnover > 1e6 ? (turnover/1e8).toFixed(1) : turnover.toFixed(1);
    candidates.push({ name, code, price, chg, volRatio, turnover: turnDisplay, pe: Math.round(pe || 0), amt: Math.round(amt*10)/10, score: Math.round(score*10)/10 });
  }
  candidates.sort(function(a,b){ return b.score - a.score; });
  return candidates.slice(0, 10);
}

async function main() {
  let all = [];
  for (let page = 1; page <= 5; page++) {
    const { stocks } = await fetchStocks(page, 100);
    all = all.concat(stocks);
    if (stocks.length < 100) break;
  }
  const picks = pick(all);
  const result = { time: new Date().toLocaleString('zh-CN', {timeZone:'Asia/Shanghai'}), totalStocks: all.length, picks };
  fs.writeFileSync('quant_picks.json', JSON.stringify(result, null, 2), 'utf8');
  console.log('scan ' + all.length + ' picked ' + picks.length);
}

main().catch(function(e){ console.error(e+" "+e.stack); process.exit(1); });

File: .github/scripts/fetch_quant.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""A股量化选股 - 多因子策略"""
import json, subprocess, urllib.request, urllib.error, ssl, sys

_CTX = ssl.create_default_context()
_CTX.check_hostname = False; _CTX.verify_mode = ssl.CERT_NONE

def curl(url):
    try:
        r = subprocess.run(['timeout','10','curl','-sL',url,'-A','Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36','--connect-timeout','8','--max-time','10','-o','-','-w',''], capture_output=True, timeout=12, text=True)
        return r.stdout
    except: return ''

def fallback(url):
    try:
        r = urllib.request.urlopen(urllib.request.Request(url, headers={'User-Agent':'Mozilla/5.0'}), timeout=10, context=_CTX)
        b = r.read()
        try: return b.decode('utf-8')
        except: return b.decode('gbk','replace')
    except: return ''

def get(url):
    h = curl(url)
    if not h or len(h) < 50:
        h = fallback(url)
    return h

def pick(stocks):
    c = []
    for s in stocks:
        name = s.get('f14','')
        code = s.get('f12','')
        if 'ST' in name or '退' in name: continue
        chg = s.get('f3',0) or 0
        if chg < 1.0 or chg > 18.0: continue
        vr = s.get('f23',0) or 0
        if vr < 0.8: continue
        to = s.get('f38',0) or 0
        if to < 0.5: continue
        pe = s.get('f115',0) or 0
        if pe > 500: continue
        amt = (s.get('f62',0) or 0) / 1e8
        cc = (s.get('f21',1) or 1) / 1e10
        rt = to/1e8 if to > 100 else to
        score = 0
        score += min(chg/5, 1.5)*20
        score += min(vr/3, 1.5)*20
        score += min(rt/5, 1.0)*15
        score += min(amt/5, 1.0)*15
        score += min(cc/30, 1.0)*15
        if 10 < pe < 50: score += 15
        c.append({'name':name,'code':code,'chg':chg,'volRatio':vr,'turnover':'%.1f'%rt,'pe':int(pe),'amt':round(amt,1),'score':round(score,1)})
    c.sort(key=lambda x:-x['score'])
    return c[:10]

def main():
    all_ = []
    for page in [1,2,3,4,5]:
        url = 'https://push2.eastmoney.com/api/qt/clist/get?pn=%d&pz=100&po=1&np=1&ut=bd1d9ddb04089700cf9c27f6f7426281&fltt=2&invt=2&fid=f62&fs=m:0+t:6,m:0+t:80,m:1+t:2,m:1+t:23&fields=f2,f3,f4,f5,f6,f7,f8,f9,f10,f12,f14,f15,f16,f17,f18,f20,f21,f23,f25,f37,f38,f62,f115,f152,f168,f169,f170,f171' % page
        h = get(url)
        try:
            items = json.loads(h).get('data',{}).get('diff',[])
            all_.extend(items)
        except: pass
        if len(items) < 100: break
    picks = pick(all_)
    with open('quant_picks.json','w',encoding='utf-8') as f:
        json.dump({'picks':picks}, f, ensure_ascii=False)
    print('scan %d picked %d' % (len(all_), len(picks)))

if __name__ == '__main__':
    main()
